src/integrations/feed/clients.ts

```diff
@@ -0,0 +1,167 @@
+import { FeedConfig, FeedResponse, IFeedClient } from './types';
+import fetch, { RequestInit } from 'node-fetch';
+import { fetchOptions as globalFetchOptions } from '../../http';
+import pRetry, { AbortError } from 'p-retry';
+import { IORedisPool } from '@dailydotdev/ts-ioredis-pool';
+import { Context } from '../../Context';
+
+const ONE_DAY_SECONDS = 24 * 60 * 60;
+
+/**
+ * Naive implementation of a feed client that fetches the feed from the feed service
+ */
+export class FeedClient implements IFeedClient {
+  private readonly url: string;
+  private readonly fetchOptions: RequestInit;
+
+  constructor(
+    url = process.env.INTERNAL_FEED,
+    fetchOptions: RequestInit = globalFetchOptions,
+  ) {
+    this.url = url;
+    this.fetchOptions = fetchOptions;
+  }
+
+  private async internalFetchFeed(config: FeedConfig): Promise<{
+    data: { post_id: string; metadata: Record<string, string> }[];
+  }> {
+    const res = await fetch(this.url, {
+      ...this.fetchOptions,
+      method: 'POST',
+      body: JSON.stringify(config),
+    });
+    if (res.status >= 200 && res.status < 300) {
+      const bodyText = await res.text();
+      return JSON.parse(bodyText);
+    }
+    if (res.status < 500) {
+      throw new AbortError(`feed service request is invalid: ${res.status}`);
+    }
+    throw new Error(`unexpecetd response from feed service: ${res.status}`);
+  }
+
+  async fetchFeed(ctx, feedId, config): Promise<FeedResponse> {
+    const res = await pRetry(() => this.internalFetchFeed(config), {
+      retries: 5,
+    });
+    if (!res.data.length) {
```

Comment:
Could it technically not even return data?
Not sure we need to optionally chain it maybe?

Reply:
Good catch!